File: backend/app/schemas/rag.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class RagScopeState(BaseModel):
    scope: Literal["paper", "code", "trace"]
    status: Literal["pending", "building", "ready", "failed"]
    chunk_count: int
    embedder: str
    model: str
    dimensions: int
    error: str | None = None
    built_at: datetime | None = None


class RagIndexStatusRead(BaseModel):
    enabled: bool
    embedder: str
    model: str
    scopes: list[RagScopeState]


class RagRebuildRequest(BaseModel):
    scope: Literal["paper", "code", "trace"] | None = None
    # Rebuild even when the stored generation already matches the source.
    force: bool = False


class RagScopeResult(BaseModel):
    # ``build_index`` returns extra keys (``reused``, ``reason``) depending on the path taken.
    model_config = ConfigDict(extra="allow")

    scope: str
    status: str
    chunk_count: int


class RagRebuildResult(BaseModel):
    results: list[RagScopeResult]


class RagSearchResult(BaseModel):
    ok: bool
    items: list[dict[str, Any]] = Field(default_factory=list)
    query: str = ""
    scope: str = ""
    embedder: str = ""
    reason: str | None = None
    searched: int = 0

File: frontend/src/types/rag.ts
export type RagScope = 'paper' | 'code' | 'trace'
export type RagIndexStatus = 'pending' | 'building' | 'ready' | 'failed'

export interface RagScopeState {
  scope: RagScope
  status: RagIndexStatus
  chunk_count: number
  embedder: string
  model: string
  dimensions: number
  error: string | null
  built_at: string | null
}

export interface RagIndexStatusRead {
  enabled: boolean
  embedder: string
  model: string
  scopes: RagScopeState[]
}

export interface RagRebuildRequest {
  /** Omit to rebuild every scope. */
  scope?: RagScope
  /** Rebuild even when the stored generation already matches the source. */
  force?: boolean
}

export interface RagScopeResult {
  scope: string
  status: string
  chunk_count: number
  reused?: boolean
  reason?: string
}

export interface RagRebuildResult {
  results: RagScopeResult[]
}

/** One retrieval hit. Extra keys depend on the scope (page/section, path/lines, verdict). */
export interface RagSearchItem {
  ref: string
  score: number
  text: string
  [key: string]: unknown
}

export interface RagSearchResult {
  ok: boolean
  items: RagSearchItem[]
  query: string
  scope: string
  embedder: string
  reason: string | null
  searched: number
}
